In [83]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn import linear_model
from sklearn.linear_model import LogisticRegression

from sklearn import svm
from sklearn.svm import SVC, LinearSVC

from sklearn import ensemble
from sklearn.neighbors import KNeighborsClassifier

from sklearn import tree
from sklearn.tree import DecisionTreeClassifier

from sklearn.linear_model import SGDClassifier

from sklearn import naive_bayes
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score

import warnings
warnings.filterwarnings('ignore')
warnings.simplefilter('ignore')


In [84]:
train = pd.read_csv('data/train.csv')
test = pd.read_csv('data/test.csv')

In [85]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [86]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

## Performing Data Cleaning and Analysis
<!-- #### 1. Understanding meaning of each column: -->
<!-- <br>Data Dictionary: -->
<br>**Variable        Description**</br>
1. Survived	- Survived (1) or died (0)
2. Pclass -	Passenger’s class (1 = 1st, 2 = 2nd, 3 = 3rd)
3. Name	- Passenger’s name
4. Sex -	Passenger’s sex
5. Age	- Passenger’s age
6. SibSp -	Number of siblings/spouses aboard
7. Parch -	Number of parents/children aboard (Some children travelled only with a nanny, therefore parch=0 for them.)
8. Ticket -	Ticket number
9. Fare -	Fare
10. Cabin -	Cabin
11. Embarked -	Port of embarkation (C = Cherbourg, Q = Queenstown, S = Southampton)


In [87]:
"""
Deleting the columns not required for determining the survival of a person
"""

del train['PassengerId']
del train['Ticket']
del train['Fare']
del train['Cabin']
del train['Name']


In [88]:
train.head()

Survived  Pclass     Sex   Age  SibSp  Parch Embarked
0         0       3    male  22.0      1      0        S
1         1       1  female  38.0      1      0        C
2         1       3  female  26.0      0      0        S
3         1       1  female  35.0      1      0        S
4         0       3    male  35.0      0      0        S

In [89]:
train.shape

(891, 7)

In [90]:
train.describe()

Survived      Pclass         Age       SibSp       Parch
count  891.000000  891.000000  714.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594
std      0.486592    0.836071   14.526497    1.102743    0.806057
min      0.000000    1.000000    0.420000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000
50%      0.000000    3.000000   28.000000    0.000000    0.000000
75%      1.000000    3.000000   38.000000    1.000000    0.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000

In [91]:
train.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Embarked      2
dtype: int64

In [92]:
"""Since there are ONLY 2 rows whose Embarked data is not known,
therefore we can neglect those 2 rows as they will not make much of a difference"""


"""
Similarly, deleting the columns not required for determining the survival of a person for the data stored in test.csv
"""

del test['Ticket']
del test['Fare']
del test['Cabin']
del test['Name']

test.head()

PassengerId  Pclass     Sex   Age  SibSp  Parch Embarked
0          892       3    male  34.5      0      0        Q
1          893       3  female  47.0      1      0        S
2          894       2    male  62.0      0      0        Q
3          895       3    male  27.0      0      0        S
4          896       3  female  22.0      1      1        S

In [93]:
test.describe()

PassengerId      Pclass         Age       SibSp       Parch
count   418.000000  418.000000  332.000000  418.000000  418.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344
std     120.810458    0.841838   14.181209    0.896760    0.981429
min     892.000000    1.000000    0.170000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000
50%    1100.500000    3.000000   27.000000    0.000000    0.000000
75%    1204.750000    3.000000   39.000000    1.000000    0.000000
max    1309.000000    3.000000   76.000000    8.000000    9.000000

In [94]:
test.isnull().sum()

PassengerId     0
Pclass          0
Sex             0
Age            86
SibSp           0
Parch           0
Embarked        0
dtype: int64

In [95]:
#We will create a new column called "Gender" and
#fill it with values 1 ,2 based on the values of sex column in which male = 1 and female = 2

def getNum(str):
    if str == 'male':
        return '1'
    if str == 'female':
        return '2'

train['Gender'] = train['Sex'].apply(getNum)
print(train.head())

print("#"*50)

test['Gender'] = test['Sex'].apply(getNum)
test.head()

   Survived  Pclass     Sex   Age  SibSp  Parch Embarked Gender
0         0       3    male  22.0      1      0        S      1
1         1       1  female  38.0      1      0        C      2
2         1       3  female  26.0      0      0        S      2
3         1       1  female  35.0      1      0        S      2
4         0       3    male  35.0      0      0        S      1
##################################################


PassengerId  Pclass     Sex   Age  SibSp  Parch Embarked Gender
0          892       3    male  34.5      0      0        Q      1
1          893       3  female  47.0      1      0        S      2
2          894       2    male  62.0      0      0        Q      1
3          895       3    male  27.0      0      0        S      1
4          896       3  female  22.0      1      1        S      2

In [96]:
del train['Sex']
del test['Sex']

## Probability of Survived to Pclass, Gender, SibSp and Parch


In [102]:
train[['Pclass', 'Survived']].groupby(['Pclass']).mean().sort_values(by='Survived', ascending=True)

Survived
Pclass          
3       0.242363
2       0.472826
1       0.629630

### Insight:
    It was revealed that passenger in the 1st have the higher survival passenger with 62.96% compare to other class

In [103]:

train[['Gender', 'Survived']].groupby(['Gender']).mean().sort_values(by='Survived', ascending=True)

Survived
Gender          
1       0.188908
2       0.742038

## Insight
    Female which is represented by 2 have the higer percent of survival with 74.20 % compare to their male counterpart

In [104]:
train[['Parch', 'Survived']].groupby(['Parch']).mean().sort_values(by='Survived', ascending=True)

Survived
Parch          
4      0.000000
6      0.000000
5      0.200000
0      0.343658
2      0.500000
1      0.550847
3      0.600000

## Insight:
    Passenger in which parent and children are 3 have 60% of survival 

In [105]:
train[['SibSp', 'Survived']].groupby(['SibSp']).mean().sort_values(by='Survived', ascending=True)

Survived
SibSp          
5      0.000000
8      0.000000
4      0.166667
3      0.250000
0      0.345395
2      0.464286
1      0.535885

## Insight:
    Passenger 

## Analyzing Data by Visualization

Inorder to understand who would have had a better probability of survival, we should visualize the patients who survived based on age, passenger class and etc.


In [ ]:
sns.countplot(train['Survived'])

In [ ]:
age_hist = sns.FacetGrid(train, col='Survived')
age_hist.map(plt.hist, 'Age')
age_hist.set_ylabels('Number')